### PR TITLE
Add AWS::Serverless::Function.Auth

### PR DESCRIPTION
**Pull request by scraymondjr**
*Issue #, if available:*

*Description of changes:*
Updates sam schema file with description of [Function Auth Object](https://github.com/aws/serverless-application-model/blob/master/versions/2016-10-31.md#function-auth-object) and generates the go code for the change.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: generate/sam-2016-10-31.json [hunk 1]
```diff
@@ -783,6 +783,124 @@
...
+                },
+                "ResourcePolicy": {
+                    "Documentation": "https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#function-auth-object",
+                    "Required": false,
+                    "Type": "AuthResourcePolicy",
+                    "UpdateType": "Immutable"
+                }
+            }
+        },
+        "AWS::Serverless::Function.AuthResourcePolicy": {
```
Comment (scraymondjr) [inline generate/sam-2016-10-31.json:825]: `ResourcePolicy` is defined in the aws docs [here](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-resourcepolicystatement.html) but not in awslabs doc. Not sure if desirable to update the `ResourcePolicy` doc to the aws doc or keep it to the same page as the rest of the types.

**Comment by PaulMaddox** [comment]
Hey @scraymondjr, this looks great. Thank you for the contribution.

Would you mind adding a test that covers all of the new possible parameters? This will help catch regressions in the future. 

Thanks

**Comment by scraymondjr** [comment]
Thanks, @PaulMaddox - added tests [here](https://github.com/awslabs/goformation/pull/373/files#diff-401b5b12739be3ae9fd2711d8490522301c12181c2424e45b027403fa1d97d8fR839).

**Comment by PaulMaddox** [comment]
Thanks adding the tests 👍 

I'll get this merged early next week.

**Pull request merged**